File: src/nylon/parser/parsers/NumberParser.java
package nylon.parser.parsers;

import nylon.NylonObject;
import nylon.nylonobjects.NylonDouble;
import nylon.nylonobjects.NylonFunction;
import nylon.nylonobjects.NylonLong;
import nylon.parser.CharIterator;
import nylon.parser.NylonParser;
import nylon.parser.Parser;

import java.util.ArrayList;
import java.util.Stack;

/**
 * Created by Aedan Smith.
 */

public class NumberParser {
    public static void addTo(ArrayList<Parser> parsers) {
        for (int i = '0'; i <= '9'; i++) {
            parsers.set(i, NumberParser::parse);
        }
    }

    public static NylonFunction parse(CharIterator in, NylonParser nylonParser) {
        boolean isDouble = false;
        String s = "";
        while (in.hasNext() && ((in.peek() >= '0' && in.peek() <= '9') || in.peek() == '.' || in.peek() == 'E')) {
            if (in.peek() == '.' || in.peek() == 'E')
                isDouble = true;
            s += in.next();
        }
        if (isDouble) {
            double d = Double.parseDouble(s);

            return new NylonFunction("PushNylonDouble(\"" + d + "\")") {
                @Override
                public void applyImpl(Stack<NylonObject> stack) {
                    NylonDouble nd = new NylonDouble(d);
                    stack.push(nd);
                }

                @Override
                public String toString() {
                    return id;
                }
            };
        } else {
            long l = Long.parseLong(s);

            return new NylonFunction("PushNylonLong(\"" + l + "\")") {
                @Override
                public void applyImpl(Stack<NylonObject> stack) {
                    NylonLong nl = new NylonLong(l);
                    stack.push(nl);
                }

                @Override
                public String toString() {
                    return id;
                }
            };
        }
    }
}

File: src/nylon/parser/parsers/NumberParser.kt
package nylon.parser.parsers

import nylon.NylonFunction
import nylon.createProvider
import nylon.nylonobjects.DoubleObject
import nylon.nylonobjects.LongObject
import nylon.parser.CharIterator
import nylon.parser.NylonParser
import nylon.parser.Parser
import nylon.parser.ParserBuilder
import java.util.*

/**
 * Created by Aedan Smith.
 */

class NumberParserBuilder : ParserBuilder {
    override fun accept(parsers: ArrayList<Parser>) {
        for (i in '0'..'9') {
            parsers[i.toInt()] = NumberParser()
        }
    }
}

class NumberParser : Parser {
    override fun apply(src: CharIterator, parser: NylonParser): NylonFunction? {
        var num = ""
        var isDouble = false
        while (src.hasNext() && (src.peek() in '0'..'9' || src.peek() == '.')) {
            if (src.peek() == '.')
                isDouble = true
            num += src.next()
        }
        if (isDouble) {
            return createProvider(DoubleObject(num.toDouble()))
        } else {
            return createProvider(LongObject(num.toLong()))
        }
    }
}
